src/software/ai/hl/stp/action/intercept_ball_action.cpp

```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
+        bool ball_very_close_to_robot_dribbler =
+            distance(ball.position(),
+                     robot->position() + Vector::createFromAngle(robot->orientation())
+                                             .normalize(DIST_TO_FRONT_OF_ROBOT_METERS)) <
+            0.2 + BALL_MAX_RADIUS_METERS;
+        Vector dribbler_tangential_velocity =
+            Vector::createFromAngle(robot->orientation())
+                .perpendicular()
+                .normalize((DIST_TO_FRONT_OF_ROBOT_METERS + BALL_MAX_RADIUS_METERS) *
+                           robot->angularVelocity().toRadians());
+        Vector expected_ball_velocity = robot->velocity() + dribbler_tangential_velocity;
+        bool velocity_magnitudes_similar =
+            std::fabs(expected_ball_velocity.length() - ball.velocity().length()) < 0.05;
+        bool velocity_angles_similar =
+            expected_ball_velocity.orientation().minDiff(ball.velocity().orientation()) <
+            Angle::fromDegrees(10);
+        bool ball_velocity_similar_to_robot =
+            velocity_angles_similar && velocity_magnitudes_similar;
+        intercept_done =
+            ball_very_close_to_robot_dribbler && ball_velocity_similar_to_robot;
+    } while (!intercept_done);
+
+    // Stop while keeping the ball in the dribbler
     do
     {
-        Point closest_point = closestPointOnLine(
-            robot->position(), Line(ball.position(), ball.position() + ball.velocity()));
-        bool point_in_front_of_ball =
-            acuteAngle(ball.velocity(), closest_point - ball.position()) <
-            Angle::quarter();
+        auto face_ball_orientation = (ball.position() - robot->position()).orientation();
 
-        Duration ball_time_to_position;
-        bool ball_not_moving = false;
-
-        if (ball.velocity().length() != 0)
-        {
-            ball_time_to_position = Duration::fromSeconds(
-                distance(closest_point, ball.position()) / (ball.velocity().length()));
-        }
-        else
-        {
-            ball_not_moving = true;
-        }
+        yield(std::make_unique<MoveIntent>(robot->id(), robot->position(),
+                                           face_ball_orientation, 0, 0,
+                                           DribblerEnable::ON, MoveType::NORMAL,
+                                           AutokickType::NONE, BallCollisionType::ALLOW));
+    } while (robot->velocity().length() > 0.03);
+}
 
+void InterceptBallAction::interceptFastBall(IntentCoroutine::push_type& yield)
+{
+    // Find the first point where the robot can get to before the ball
+    Point intercept_position = ball.position();
+    while (contains(field.fieldLines(), intercept_position))
+    {
+        Duration ball_time_to_position =
+            Duration::fromSeconds(distance(intercept_position, ball.position()) /
+                                  (ball.velocity().length() + 1e-6));
```

Comment:
can we change 1e-6 to a constant or avoid it?

Reply:
avoided it since this case is only called when the ball is moving fast